rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,592 @@
+            GetConnectionOrCreate::AddConnection(add) => {
+                let conn = create_conn().await?;
+                Ok(add(conn))
+            }
+        }
+    }
+
+    /// Get connection from pool or return fn to add a new one. See [`GetConnectionOrCreate`] for more info
+    pub async fn get_connection_or_create_cb(
+        &self,
+        id: &ConnId,
+    ) -> Result<
+        GetConnectionOrCreate<C, impl FnOnce(C) -> LeasedConnection<C, ConnId>, ConnId>,
+        OpaqueError,
+    > {
+        let active_permit = self
+            .inner
+            .active_slots
+            .clone()
+            .acquire_owned()
+            .await
+            .context("failed to acquire active slot permit")?;
+
+        let active_slot = ActiveSlot(active_permit);
+        let pool = Arc::<PoolInner<C, ConnId>>::downgrade(&self.inner);
+
+        // Check if we can reuse stored connection
+        let pooled_conn = {
+            let mut connections = self.inner.connections.lock();
+            connections
+                .iter()
+                .position(|stored| &stored.id == id)
+                .map(|idx| connections.remove(idx))
+                .flatten()
+        };
+
+        if pooled_conn.is_some() {
+            let leased_conn = LeasedConnection {
+                _slot: active_slot,
+                pooled_conn,
+                pool,
+            };
+            return Ok(GetConnectionOrCreate::LeasedConnection(leased_conn));
+        };
+
+        // If we have an active slot we should always be able to get a pool slot. Unless our pool
+        // is totally full, and in that case we just need to drop one connection and use that slot instead
+        let pool_slot = match self.inner.total_slots.clone().try_acquire_owned() {
+            Ok(pool_permit) => PoolSlot(pool_permit),
+            Err(_) => {
+                let pooled_conn = self.inner.connections.lock().pop_back().context(
+                    "connections vec cannot be empty if total slots doesn't have permit available",
+                )?;
+                pooled_conn.slot
+            }
+        };
+
+        Ok(GetConnectionOrCreate::AddConnection(move |conn: C| {
+            LeasedConnection {
+                _slot: active_slot,
+                pool,
+                pooled_conn: Some(PooledConnection {
+                    conn,
+                    id: id.clone(),
+                    slot: pool_slot,
+                }),
+            }
+        }))
+    }
+}
+
+/// [`ReqToConnId`] is used to convert a `Request` to a connection ID. These IDs are
+/// not unique and multiple connections can have the same ID. IDs are used to filter
+/// which connections can be used for a specific Request in a way that is indepent of
+/// what a Request is.
+pub trait ReqToConnId<Request>: Sized + Send + Sync + 'static
+where
+    Request: Send + 'static,
+{
+    type ConnId: Send + Sync + PartialEq + Clone;
```

Comment:
You porbably want to require `'static` as well, at least in the current version of rama I think it will help